### ISSUE TITLE
untrusted certificates should report non-conformity with Apple guidelines for validity period and required extensions and key sizes

### ISSUE THREAD

**Issue by jtkohl** [certificate-helper/TLS-Inspector#179]
**Affected Version:**

2.0.3

**Is this a Test Flight version or the App Store version?**

App Store version

**Device and iOS Version:**

iPad Air 2, iOS 13.3.1

**What steps will reproduce the problem?**

1.  Create your own certificate authority root signing certificate
2. Use the CA to sign a subject certificate for your own web site
3. Install the CA root certificate as a trusted root CA on iOS (install a profile with the root CA public key, and then mark it as trusted in the certificate trust settings ("enable full trust for root certificates")
4. Install the subject certificate and its chain in your web server
5. connect to the web server with TLS inspector

**What is the expected output?**

TLS inspector should show the site as trusted, like Safari shows.

**What do you see instead?**

TLS says the site is not trusted

**Please provide any additional information below.**

Previous issue #82 claims to have fixed it, but not for me.  Possibly something new in iOS 13?

**Comment by ecnepsnai** [comment]
Hi @jtkohl I'm having trouble reproducing this issue on my end. I have a locally trusted CA on my phone and inspecting sites shows up as "Locally Trusted"

![local](https://user-images.githubusercontent.com/1607109/76691430-e8842500-6607-11ea-82b7-7038f97f8db6.png)

If you tap on the (i) where it says Untrusted it will give a reason why its failing, what is that reason?

If you are comfortable could you also enable debug logging and inspect the site and submit those logs to me.

**Comment by jtkohl** [comment]
Well, I did find earlier tonight that I had forgotten to include the SAN on one of my servers' certificates, so I fixed that up, but that didn't solve the complaint from TLS inspector.
macOS Google chrome and Firefox are happy with the certificates, given that the root CA is in my macOS keychain of trusted CAs.

![bad-trust](https://user-images.githubusercontent.com/52104389/76717838-3897e000-670b-11ea-9346-38a31254f565.jpg)
and the inspector log:
[TLS Inspector.log](https://github.com/certificate-helper/TLS-Inspector/files/4335737/TLS.Inspector.log)

and an openssl debug/showcerts transcript from:
openssl s_client -debug -connect lede.lan:443 -showcerts -CAfile caCert.pem
(caCert.pem has the same certificate as shown in the openssl chain)
[openssl.log](https://github.com/certificate-helper/TLS-Inspector/files/4335744/openssl.log)

**Comment by ecnepsnai** [comment]
Thank you for the openssl output @jtkohl, that proved to be exactly what I needed to reproduce the issue on my end.

**Comment by ecnepsnai** [comment]
@jtkohl Actually disregard my prior comment, I'm still having issues reproducing the problem.

What web server are you using? Are you able to provide the relevant ssl configuration? One thing I noticed in the TLS Inspector log is that it's having trouble showing the protocol version and ciphersuite:

```
Connected to 'lede.lan' (192.168.1.1), Protocol version: (null), Ciphersuite: (null)
```

**Comment by jtkohl** [comment]
I'm using OpenWRT 19.07.2, with uhttpd.  My server is set up with the two certificates in the chain stored as PEM format.

Here's the list of installed packages, sorry I'm not sure exactly which SSL variant is in use for uhttpd.

    ath10k-firmware-qca9984-ct - 2019-10-03-d622d160-1
    avahi-dbus-daemon - 0.7-3
    avahi-utils - 0.7-3
    base-files - 204.2-r10947-65030d81f3
    block-mount - 2020-01-18-189b41b6-1
    busybox - 1.30.1-5
    ca-bundle - 20190110-2
    ca-certificates - 20190110-2
    cgi-io - 19
    dbus - 1.12.12-1
    ddns-scripts - 2.7.8-12
    diffutils - 3.7-2
    dnsmasq - 2.80-15
    dropbear - 2019.78-2
    e2fsprogs - 1.44.5-2
    f2fs-tools - 1.12.0-3
    f2fsck - 1.12.0-3
    firewall - 2019-11-22-8174814a-1
    fstools - 2020-01-18-189b41b6-1
    fwtool - 2
    gdisk - 1.0.4-2
    getrandom - 2019-06-16-4df34a4d-3
    hostapd-common - 2019-08-08-ca8c2bd2-2
    ip-tiny - 5.0.0-2.1
    ip6tables - 1.8.3-1
    iptables - 1.8.3-1
    iptables-mod-conntrack-extra - 1.8.3-1
    iptables-mod-extra - 1.8.3-1
    iptables-mod-ipopt - 1.8.3-1
    iptables-mod-ipsec - 1.8.3-1
    iptables-mod-nat-extra - 1.8.3-1
    iptables-mod-nflog - 1.8.3-1
    irqbalance - 1.6.0-7
    iw-full - 5.0.1-1
    jshn - 2020-02-27-7da66430-1
    jsonfilter - 2018-02-04-c7e938d6-1
    kernel - 4.14.171-1-0894164cab0effc42201a29fec8ce33f
    kmod-ata-ahci - 4.14.171-1
    kmod-ata-ahci-platform - 4.14.171-1
    kmod-ata-core - 4.14.171-1
    kmod-ath - 4.14.171+4.19.98-1-1
    kmod-ath10k-ct - 4.14.171+2019-09-09-5e8cd86f-1
    kmod-cfg80211 - 4.14.171+4.19.98-1-1
    kmod-crypto-acompress - 4.14.171-1
    kmod-crypto-aead - 4.14.171-1
    kmod-crypto-authenc - 4.14.171-1
    kmod-crypto-cbc - 4.14.171-1
    kmod-crypto-crc32 - 4.14.171-1
    kmod-crypto-crc32c - 4.14.171-1
    kmod-crypto-deflate - 4.14.171-1
    kmod-crypto-des - 4.14.171-1
    kmod-crypto-echainiv - 4.14.171-1
    kmod-crypto-hash - 4.14.171-1
    kmod-crypto-hmac - 4.14.171-1
    kmod-crypto-iv - 4.14.171-1
    kmod-crypto-manager - 4.14.171-1
    kmod-crypto-md5 - 4.14.171-1
    kmod-crypto-null - 4.14.171-1
    kmod-crypto-pcompress - 4.14.171-1
    kmod-crypto-rng - 4.14.171-1
    kmod-crypto-sha1 - 4.14.171-1
    kmod-crypto-sha256 - 4.14.171-1
    kmod-crypto-user - 4.14.171-1
    kmod-crypto-wq - 4.14.171-1
    kmod-fs-ext4 - 4.14.171-1
    kmod-fs-f2fs - 4.14.171-1
    kmod-gpio-button-hotplug - 4.14.171-3
    kmod-hwmon-core - 4.14.171-1
    kmod-ifb - 4.14.171-1
    kmod-ip6tables - 4.14.171-1
    kmod-ipsec - 4.14.171-1
    kmod-ipsec4 - 4.14.171-1
    kmod-ipsec6 - 4.14.171-1
    kmod-ipt-conntrack - 4.14.171-1
    kmod-ipt-conntrack-extra - 4.14.171-1
    kmod-ipt-core - 4.14.171-1
    kmod-ipt-extra - 4.14.171-1
    kmod-ipt-ipopt - 4.14.171-1
    kmod-ipt-ipsec - 4.14.171-1
    kmod-ipt-nat - 4.14.171-1
    kmod-ipt-nat-extra - 4.14.171-1
    kmod-ipt-nflog - 4.14.171-1
    kmod-ipt-offload - 4.14.171-1
    kmod-ipt-raw - 4.14.171-1
    kmod-iptunnel4 - 4.14.171-1
    kmod-iptunnel6 - 4.14.171-1
    kmod-leds-gpio - 4.14.171-1
    kmod-lib-crc-ccitt - 4.14.171-1
    kmod-lib-crc16 - 4.14.171-1
    kmod-lib-zlib-deflate - 4.14.171-1
    kmod-lib-zlib-inflate - 4.14.171-1
    kmod-mac80211 - 4.14.171+4.19.98-1-1
    kmod-nf-conntrack - 4.14.171-1
    kmod-nf-conntrack6 - 4.14.171-1
    kmod-nf-flow - 4.14.171-1
    kmod-nf-ipt - 4.14.171-1
    kmod-nf-ipt6 - 4.14.171-1
    kmod-nf-nat - 4.14.171-1
    kmod-nf-reject - 4.14.171-1
    kmod-nf-reject6 - 4.14.171-1
    kmod-nfnetlink - 4.14.171-1
    kmod-nfnetlink-log - 4.14.171-1
    kmod-nls-base - 4.14.171-1
    kmod-ppp - 4.14.171-1
    kmod-pppoe - 4.14.171-1
    kmod-pppox - 4.14.171-1
    kmod-sched-cake - 4.14.171+2019-03-12-057c7388-1
    kmod-sched-core - 4.14.171-1
    kmod-scsi-core - 4.14.171-1
    kmod-slhc - 4.14.171-1
    kmod-tun - 4.14.171-1
    kmod-usb-core - 4.14.171-1
    kmod-usb-dwc3 - 4.14.171-1
    kmod-usb-dwc3-of-simple - 4.14.171-1
    kmod-usb-ehci - 4.14.171-1
    kmod-usb-ledtrig-usbport - 4.14.171-1
    kmod-usb-ohci - 4.14.171-1
    kmod-usb-phy-qcom-dwc3 - 4.14.171-1
    kmod-usb-storage - 4.14.171-1
    kmod-usb-storage-uas - 4.14.171-1
    kmod-usb2 - 4.14.171-1
    kmod-usb3 - 4.14.171-1
    libattr - 2.4.48-1
    libavahi-client - 0.7-3
    libavahi-dbus-support - 0.7-3
    libblkid1 - 2.34-1
    libblobmsg-json - 2020-02-27-7da66430-1
    libc - 1.1.24-2
    libcap - 2.27-1
    libcomerr0 - 1.44.5-2
    libcurl4 - 7.66.0-1
    libdaemon - 0.14-5
    libdb47 - 4.7.25.4.NC-5
    libdbus - 1.12.12-1
    libelf1 - 0.177-1
    libevent2-core7 - 2.1.11-1
    libevent2-pthreads7 - 2.1.11-1
    libexpat - 2.2.7-1
    libext2fs2 - 1.44.5-2
    libf2fs6 - 1.12.0-3
    libgcc1 - 7.5.0-2
    libgcrypt - 1.8.5-1
    libgdbm - 1.11-1
    libgmp10 - 6.1.2-2
    libgpg-error - 1.36-3
    libip4tc2 - 1.8.3-1
    libip6tc2 - 1.8.3-1
    libiwinfo-lua - 2019-10-16-07315b6f-1
    libiwinfo20181126 - 2019-10-16-07315b6f-1
    libjson-c2 - 0.12.1-3
    libjson-script - 2020-02-27-7da66430-1
    liblua5.1.5 - 5.1.5-3
    liblucihttp-lua - 2019-07-05-a34a17d5-1
    liblucihttp0 - 2019-07-05-a34a17d5-1
    liblzo2 - 2.10-2
    libmariadb - 10.2.31-2
    libmbedtls12 - 2.16.4-1
    libmnl0 - 1.0.4-2
    libncurses6 - 6.1-5
    libnl-tiny - 0.1-5
    libopenldap - 2.4.48-1
    libopenssl-conf - 1.1.1d-2
    libopenssl1.1 - 1.1.1d-2
    libpcap1 - 1.9.1-1
    libpcre - 8.43-1
    libpthread - 1.1.24-2
    librt - 1.1.24-2
    libsasl2 - 2.1.27-1
    libsqlite3-0 - 3310100-1
    libss2 - 1.44.5-2
    libubox20191228 - 2020-02-27-7da66430-1
    libubus-lua - 2019-12-27-041c9d1c-1
    libubus20191227 - 2019-12-27-041c9d1c-1
    libuci20130104 - 2019-09-01-415f9e48-3
    libuclient20160123 - 2019-05-30-3b3e368d-1
    libusb-1.0-0 - 1.0.22-2
    libustream-mbedtls20150806 - 2019-11-05-c9b66682-2
    libuuid1 - 2.34-1
    libxml2 - 2.9.9-2
    libxtables12 - 1.8.3-1
    logd - 2019-06-16-4df34a4d-3
    lua - 5.1.5-3
    luci - git-20.067.35699-cf18bdd-1
    luci-app-ddns - 2.4.9-7
    luci-app-firewall - git-20.067.35699-cf18bdd-1
    luci-app-openvpn - git-20.067.35699-cf18bdd-1
    luci-app-opkg - git-20.067.35699-cf18bdd-1
    luci-app-sqm - 1.4.0-2
    luci-app-uhttpd - git-20.067.35699-cf18bdd-1
    luci-base - git-20.067.35699-cf18bdd-1
    luci-compat - git-20.067.35699-cf18bdd-1
    luci-lib-ip - git-20.067.35699-cf18bdd-1
    luci-lib-ipkg - git-20.067.35699-cf18bdd-1
    luci-lib-jsonc - git-20.067.35699-cf18bdd-1
    luci-lib-nixio - git-20.067.35699-cf18bdd-1
    luci-mod-admin-full - git-20.067.35699-cf18bdd-1
    luci-mod-network - git-20.067.35699-cf18bdd-1
    luci-mod-status - git-20.067.35699-cf18bdd-1
    luci-mod-system - git-20.067.35699-cf18bdd-1
    luci-proto-ipv6 - git-20.067.35699-cf18bdd-1
    luci-proto-ppp - git-20.067.35699-cf18bdd-1
    luci-ssl - git-20.067.35699-cf18bdd-1
    luci-theme-bootstrap - git-20.067.35699-cf18bdd-1
    mariadb-common - 10.2.31-2
    mdns-utils - 878.200.35-1
    mdnsd - 878.200.35-1
    mdnsresponder - 878.200.35-1
    mkf2fs - 1.12.0-3
    mtd - 24
    nano - 4.8-1
    netatalk - 3.1.12-1
    netifd - 2019-08-05-5e02f944-1
    ntp-utils - 4.2.8p13-3
    odhcp6c - 2019-01-11-e199804b-16
    odhcpd-ipv6only - 2019-12-16-e53fec89-3
    openssl-util - 1.1.1d-2
    openvpn-openssl - 2.4.7-2
    openwrt-keyring - 2019-07-25-8080ef34-1
    opkg - 2020-01-25-c09fe209-1
    ppp - 2.4.7.git-2019-05-25-3
    ppp-mod-pppoe - 2.4.7.git-2019-05-25-3
    procd - 2020-01-24-31e4b2df-1
    px5g-mbedtls - 9
    rpcd - 2019-12-10-aaa08366-2
    rpcd-mod-file - 2019-12-10-aaa08366-2
    rpcd-mod-iwinfo - 2019-12-10-aaa08366-2
    rpcd-mod-luci - 20191114
    rpcd-mod-rrdns - 20170710
    shadow-common - 4.8.1-1
    shadow-groupadd - 4.8.1-1
    shadow-groupmod - 4.8.1-1
    shadow-useradd - 4.8.1-1
    shadow-usermod - 4.8.1-1
    sqm-scripts - 1.4.0-2
    strongswan - 5.8.2-1
    strongswan-charon - 5.8.2-1
    strongswan-charon-cmd - 5.8.2-1
    strongswan-full - 5.8.2-1
    strongswan-ipsec - 5.8.2-1
    strongswan-libtls - 5.8.2-1
    strongswan-mod-addrblock - 5.8.2-1
    strongswan-mod-aes - 5.8.2-1
    strongswan-mod-af-alg - 5.8.2-1
    strongswan-mod-agent - 5.8.2-1
    strongswan-mod-attr - 5.8.2-1
    strongswan-mod-attr-sql - 5.8.2-1
    strongswan-mod-blowfish - 5.8.2-1
    strongswan-mod-ccm - 5.8.2-1
    strongswan-mod-cmac - 5.8.2-1
    strongswan-mod-connmark - 5.8.2-1
    strongswan-mod-constraints - 5.8.2-1
    strongswan-mod-coupling - 5.8.2-1
    strongswan-mod-ctr - 5.8.2-1
    strongswan-mod-curl - 5.8.2-1
    strongswan-mod-curve25519 - 5.8.2-1
    strongswan-mod-des - 5.8.2-1
    strongswan-mod-dhcp - 5.8.2-1
    strongswan-mod-dnskey - 5.8.2-1
    strongswan-mod-duplicheck - 5.8.2-1
    strongswan-mod-eap-identity - 5.8.2-1
    strongswan-mod-eap-md5 - 5.8.2-1
    strongswan-mod-eap-mschapv2 - 5.8.2-1
    strongswan-mod-eap-radius - 5.8.2-1
    strongswan-mod-eap-tls - 5.8.2-1
    strongswan-mod-farp - 5.8.2-1
    strongswan-mod-fips-prf - 5.8.2-1
    strongswan-mod-forecast - 5.8.2-1
    strongswan-mod-gcm - 5.8.2-1
    strongswan-mod-gcrypt - 5.8.2-1
    strongswan-mod-gmp - 5.8.2-1
    strongswan-mod-ha - 5.8.2-1
    strongswan-mod-hmac - 5.8.2-1
    strongswan-mod-kernel-netlink - 5.8.2-1
    strongswan-mod-ldap - 5.8.2-1
    strongswan-mod-led - 5.8.2-1
    strongswan-mod-load-tester - 5.8.2-1
    strongswan-mod-md4 - 5.8.2-1
    strongswan-mod-md5 - 5.8.2-1
    strongswan-mod-mysql - 5.8.2-1
    strongswan-mod-nonce - 5.8.2-1
    strongswan-mod-openssl - 5.8.2-1
    strongswan-mod-pem - 5.8.2-1
    strongswan-mod-pgp - 5.8.2-1
    strongswan-mod-pkcs1 - 5.8.2-1
    strongswan-mod-pkcs11 - 5.8.2-1
    strongswan-mod-pkcs12 - 5.8.2-1
    strongswan-mod-pkcs7 - 5.8.2-1
    strongswan-mod-pkcs8 - 5.8.2-1
    strongswan-mod-pubkey - 5.8.2-1
    strongswan-mod-random - 5.8.2-1
    strongswan-mod-rc2 - 5.8.2-1
    strongswan-mod-resolve - 5.8.2-1
    strongswan-mod-revocation - 5.8.2-1
    strongswan-mod-sha1 - 5.8.2-1
    strongswan-mod-sha2 - 5.8.2-1
    strongswan-mod-smp - 5.8.2-1
    strongswan-mod-socket-default - 5.8.2-1
    strongswan-mod-sql - 5.8.2-1
    strongswan-mod-sqlite - 5.8.2-1
    strongswan-mod-sshkey - 5.8.2-1
    strongswan-mod-stroke - 5.8.2-1
    strongswan-mod-test-vectors - 5.8.2-1
    strongswan-mod-uci - 5.8.2-1
    strongswan-mod-unity - 5.8.2-1
    strongswan-mod-updown - 5.8.2-1
    strongswan-mod-vici - 5.8.2-1
    strongswan-mod-whitelist - 5.8.2-1
    strongswan-mod-x509 - 5.8.2-1
    strongswan-mod-xauth-eap - 5.8.2-1
    strongswan-mod-xauth-generic - 5.8.2-1
    strongswan-mod-xcbc - 5.8.2-1
    strongswan-pki - 5.8.2-1
    strongswan-scepclient - 5.8.2-1
    strongswan-swanctl - 5.8.2-1
    swconfig - 12
    tc - 5.0.0-2.1
    tcpdump - 4.9.3-1
    terminfo - 6.1-5
    ubi-utils - 2.1.1-1
    uboot-envtools - 2018.03-3
    ubox - 2019-06-16-4df34a4d-3
    ubus - 2019-12-27-041c9d1c-1
    ubusd - 2019-12-27-041c9d1c-1
    uci - 2019-09-01-415f9e48-3
    uclibcxx - 0.2.5-3
    uclient-fetch - 2019-05-30-3b3e368d-1
    uhttpd - 2020-02-12-2ee323c0-1
    urandom-seed - 1.0-1
    urngd - 2020-01-21-c7f7b6b6-1
    usbutils - 007-10
    usign - 2019-08-06-5a52b379-1
    wget - 1.20.3-2
    wireless-regdb - 2019.06.03
    wpad-basic - 2019-08-08-ca8c2bd2-2
    zlib - 1.2.11-3

**Comment by jtkohl** [comment]
and the uhttpd config file on the openwrt router:
(note I didn't generate the certificates and keys on the router, I generated them on a separate machine where I have my private CA setup)

    root@LEDE:~# cat /etc/config/uhttpd
    
    config uhttpd 'main'
    	list listen_http '0.0.0.0:80'
    	list listen_http '[::]:80'
    	list listen_https '0.0.0.0:443'
    	list listen_https '[::]:443'
    	option redirect_https '1'
    	option home '/www'
    	option rfc1918_filter '1'
    	option max_requests '3'
    	option max_connections '100'
    	option cert '/etc/uhttpd.crt'
    	option key '/etc/uhttpd.key'
    	option cgi_prefix '/cgi-bin'
    	option script_timeout '60'
    	option network_timeout '30'
    	option http_keepalive '20'
    	option tcp_keepalive '1'
    	option ubus_prefix '/ubus'
    
    config cert 'defaults'
    	option days '3655'
    	option bits '2048'
    	option country 'US'
    	option state 'Massachusetts'
    	option location 'Arlington'
    	option commonname 'LEDE.lan'

**Comment by ecnepsnai** [comment]
Thanks again (seriously you're giving me so much useful info here - especially with that openssl log).

I want to highlight that the ability to identify locally trusted certificates (as opposed to those included in the systems root store) is a very important feature to TLS Inspector as it's a very common tactic used by malware on iOS.

Looking at the specific trust result that the iOS SDK is giving us for the chain indicated a "recoverable trust failure". It looks like I might be able to get more information from the system as to what specifically is failing here.

To clarify on how the logic works: we pass the certificate chain and SSL context to iOS and ask it to evaluate it. For valid system-trusted chains the system returns `unspecified`, for locally trusted chains the system returns `proceed`, these are the only two results that are considered trustworthy, the rest are considered failures. In that case, we attempt to manually determine what the trust failure is, in this case we're unable to do that and defaulting to a generic error.

So what this tells me is that:
1. The SDK available to iOS apps for certificate trust actually doesn't trust your certificate, but...
2. Safari appears to either be using a different SDK - or is doing something else and is considering the chain to actually be valid

What I'm going to do:
1. Continue trying to replicate the issue locally, I might also set up an openwrt server and see if it's specific to that platform
2. Look for ways to get more information out of the iOS SDK as to why it's not trusting the chain, this will require me to push an update to the app store and for you to update and attach new logs, but we'll cross that bridge if and when.

**Comment by ecnepsnai** [comment]
@jtkohl I believe I've identified the root cause of the issue here - your server certificate doesn't contain either the key or extended key usage extension and therefor can't (legitimately) be used for a web server.

It seems to be up to the browser to decide if it wants to ignore this or not with differing results between browsers and systems. On macOS it seems that if the root CA is trusted then it ignores it in Chrome and Safari. On iOS it seems to be the same.

I've reached out to a chromium developer for his thoughts on the matter, and will extend the validation logic to highlight these certificates.

In the meantime I suggest you regenerate your server certificate with the following usages:
- Digital Signature
- Key Encipherment
- Server Authentication
- Client Authentication

**Comment by jtkohl** [comment]
tl;dr:  certificates must meet new limitations as published by Apple.  Mine were valid too long.

May I suggest converting this to an RFE to have your code check the reasons listed in Apple's support document as a hint to the user?  :-)


Ah, thanks for helping diagnose this.  I found the missing part in my scripting that didn't select the key/exended key usage sections.
I regenerated the cert as you suggested (and also with only server-authentication, it seems unnecessary to sign this cert for client authentication since it's only going to be used on a web server), and still TLS inspector doesn't like it.

Hmm, doing some more testing, safari on iOS doesn't like either the cert without the key/extended key, or the one with those extensions.  But safari does like the old cert from January which has no SAN extension and no key usage extensions.  However, TLS Inspector doesn't like the certificate with no SAN (perfectly reasonable thing to insist upon, Chrome does for example), saying the cert is for a different host.
And when I reissue a certificate with a SAN for DNS:lede.lan and IP:192.168.1.1, and use the IP address, iOS Safari doesn't like that either.

Thinking there was something about iOS with private CAs that doesn't like my certs, I ran a web search and found:
[https://apple.stackexchange.com/questions/371725/why-does-ios-13-not-trust-my-own-root-ca](https://apple.stackexchange.com/questions/371725/why-does-ios-13-not-trust-my-own-root-ca) and [https://support.apple.com/en-us/HT210176](https://support.apple.com/en-us/HT210176) and those explain that iOS will reject long-lived server certificates.  I signed mine for 10 years to simplify my life, but iOS barfs on that duration.
(Oddly, the claim that iOS requires a SAN field doesn't match my experience with my certificate from January, with no SAN field and a 10-year validity.  That cert is accepted by iOS safari.)

When I reissued a certificate valid for 2 years, with SAN and key usage/extended key usage showing Digital Signature, Key Encipherment, TLS Web Server Authentication, all is good.

Here's an example:

    Data:
        Version: 3 (0x2)
        Serial Number: 30 (0x1e)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=US, O=Boston Pog Party, CN=bostonpog root
        Validity
            Not Before: Mar 22 12:53:39 2020 GMT
            Not After : Mar 22 12:53:39 2022 GMT
        Subject: C=US, ST=MA, O=Boston Pog Party, CN=lede.lan
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:dc:60:e6:d4:59:94:46:db:0d:03:25:06:4d:80:
                    f1:fa:e7:e1:a3:29:cd:80:a4:88:e4:40:e1:45:47:
                    86:07:6e:85:6e:52:43:56:f4:7b:6f:fa:f6:53:24:
                    a0:31:d8:f8:96:5e:b4:98:e5:f5:3d:27:9d:2a:19:
                    53:35:d1:09:05:64:f1:a2:18:8b:06:8a:f0:a5:4b:
                    b1:b5:b0:2e:5d:b5:46:16:13:e0:35:19:3a:b5:99:
                    00:ad:26:e7:35:72:c4:19:91:f9:a8:10:fb:8c:ce:
                    69:ba:d5:b1:be:44:88:4d:e6:39:6c:be:db:ad:03:
                    ff:84:9f:85:ee:a6:18:41:95:27:58:99:61:a4:dc:
                    98:45:66:36:23:8e:26:2e:53:a1:9d:28:21:f9:f1:
                    28:26:0d:07:f8:3c:fa:84:f5:c1:dc:7f:24:ef:83:
                    dd:71:4e:26:17:c4:d8:72:fc:dc:ac:79:d0:46:00:
                    4c:90:63:67:c9:bf:ed:a0:41:e0:29:49:12:73:40:
                    0d:b8:0b:31:4d:a4:5d:7c:42:e9:2b:0f:d2:1f:00:
                    31:6a:38:92:f0:7a:d8:2a:cb:28:4a:c2:60:f6:03:
                    07:8e:ff:35:10:4e:15:b2:dc:47:86:9a:a1:f2:7a:
                    74:6b:cc:9c:67:a6:3c:85:7e:8d:04:b2:ad:e0:65:
                    1f:97
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: critical
                TLS Web Server Authentication
            X509v3 Subject Key Identifier: 
                3A:9F:D3:4D:71:D1:9E:6E:77:DF:F3:D1:5D:B2:EB:12:0D:46:56:18
            X509v3 Authority Key Identifier: 
                keyid:A7:48:21:31:B7:08:CC:F1:B3:4E:8E:C0:B3:E3:63:FE:07:7D:E4:5C

            X509v3 Basic Constraints: critical
                CA:FALSE
            Netscape Comment: 
                OpenSSL Generated Certificate
            X509v3 Subject Alternative Name: 
                DNS:lede.lan
    Signature Algorithm: sha256WithRSAEncryption
         1f:90:eb:3f:ec:92:83:51:29:fd:21:16:f4:96:d6:14:bc:db:
         c6:70:f0:0a:6b:c7:5a:f1:26:47:77:6b:20:8e:c3:2b:af:8c:
         07:cb:d4:08:8e:cd:0e:93:c9:76:03:b3:80:ec:39:d2:b2:8b:
         cf:0f:58:69:87:db:fa:a1:21:22:05:f1:64:ef:23:d9:d7:e3:
         37:74:f0:69:0c:3c:93:d5:14:85:e9:2e:c9:fa:75:64:15:2c:
         82:aa:f0:54:9b:cf:59:1b:89:bf:fc:67:1f:63:9b:41:3c:d1:
         b7:c5:3f:1e:86:7a:1a:9f:6f:bc:f3:4a:ac:31:06:9a:48:98:
         ce:0a:6e:46:87:6a:fc:e9:02:0d:c2:7f:e8:ad:33:e7:82:54:
         66:42:d4:5a:db:03:bb:fc:d3:a9:24:63:74:94:35:b5:16:c1:
         3a:fd:54:26:4e:da:34:cd:a4:5b:d8:b1:d0:bf:ec:28:9e:59:
         20:78:e6:18:c8:a1:c6:33:06:56:8e:19:f1:40:82:44:66:09:
         95:ab:ee:76:c8:a8:e4:b3:40:ab:03:39:c9:1b:8b:e5:3c:2a:
         54:02:f1:83:b1:61:e1:88:4f:9f:d4:33:5c:9f:8c:be:98:31:
         4f:6a:6f:d7:35:24:68:c2:7d:7e:f0:0a:80:cb:11:56:02:cf:
         3c:21:9b:e0

**Comment by ecnepsnai** [comment]
Ah, that explains why I wasn't running into this issue - I was using test certs valid only for 90 days.

I also want to bring up the result of what Chrome does for certs missing KU/EKU for reference: https://twitter.com/ecnepsnai/status/1241606683421315074

**Comment by ecnepsnai** [comment]
**Work to do:**

- [ ] Report certificates with weak RSA keys (<2048)
- [x] Report certificates what use SHA-1 (already done)
- [x] Report certificates that don't have a DNS object in SAN (already done)
- [ ] Report leaf certificates that don't have server auth EKU
- [ ] Report leaf certificates that are issued for greater than 825 days

**Comment by ecnepsnai** [comment]
Resolved in release 2.0.4. Thank you so much for your help, @jtkohl!
